Add guard for calling `finalize` when NullSession is present

leequarella opened the pull request:
https://github.com/rails/rails/pull/41372 Added `ActiveRecord::AsynchronousQueriesTracker::NullSession` which replaced a
use of `nil` in `AsynchronousQueriesTracker`.

This commit changes the `finalize_session` method to match that change from `nil` and properly handle cases where it is called with a `NullSession` present.

leequarella committed:
Add guard for calling `finalize` on NullSession

https://github.com/rails/rails/pull/41372 Added
`ActiveRecord::AsynchronousQueriesTracker::NullSession` which replaced a
use of `nil` in `AsynchronousQueriesTracker`.  This commit changes the
`finalize_session` method to match that change from `nil` and properly
handle cases where it is called with a `NullSession` present.

rafaelfranca commented on activerecord/lib/active_record/asynchronous_queries_tracker.rb:
```diff
@@ -50,7 +50,7 @@ def start_session
 
     def finalize_session
-      @current_session&.finalize
+      @current_session.finalize if @current_session.respond_to?(:finalize)
```
The point of using a null object is to avoid conditionals. In that case the `NullSession` should implement `finalize` and the `&` should be probably removed.

leequarella replied:
That makes sense. I've added a noop `finalize` method to `NullSession`. I don't think it actually has to do anything for now?

leequarella committed:
Add `finalize` to NullSession so it quacks like a Session

eileencodes commented:
I'm not sure if we should make this change anymore. While it will raise an error if `finalize` is called on `NullSession` I'm not sure that can happen in a real application. For us to have `NullSession` at this point in the code, we must have accidentally nested calls to`run` and `complete` (something I don't think is possible in Rails). It happened to us due to our Sinatra API that mixes in Rails. I can reproduce with a test like this:

```ruby
class AsyncQueriesMiddlewareTest < ActiveRecord::TestCase
  def test_writes_should_always_clear_cache
    mw = middleware { |env|
      middleware { |env|
        p "broken"
      }.call({})
    }
    mw.call({})
  end

  def middleware(&app)
    executor = Class.new(ActiveSupport::Executor)
    ActiveRecord::AsynchronousQueriesTracker.install_executor_hooks executor
    lambda { |env| executor.wrap { app.call(env) } }
  end
end
```

At GitHub we found that in certain tests our middleware for the Executor was getting nested causing `run` and `complete` to be called twice. The main issue was that the Sinatra API wants to use the Query Cache too so we had `use ActionDispatch::Executor, ActiveSupport::Executor` in our Sinatra API. This meant that the Sinatra API was using a different executor object than Rails - which caused the issue. In tests that ran an API request that fired and active job, that resulted in `run` and `complete` getting called twice - once for Sinatra's Executor and once for Active Job's executor. Changing this to `use ActionDispatch::Executor, Rails.application.executor` fixes this. We never saw this before the async executor was added because there isn't anything `complete` calls that would raise an error in that the query cache middleware, but my investigation showed it's been nested the whole time.

eileencodes commented:
I do like removing the `&finalize` though so I like that part of the change and is a good reason to keep it. I wanted to document the investigation at GitHub and what our app is doing that's inconsistent with a standard Rails application.

leequarella committed:
Remove `finalize` from NullSession

`finalize_session` should not be able to be called if a session has not
been started, we should never have a `NullSession` when we get there. So
we can remove the guard altogether to help make it obvious if this sort
of thing does happen when it's not supposed to.

leequarella commented:
Stripped this down so it only removes the maybe operator for `.finalize`. This should help to avoid misuse of the `finalize_session` method.
